### ISSUE TITLE
Allow for multiple configurations in the same app. (sandbox/prod or different push topics).

### ISSUE THREAD

**Issue by idyll** [codedge-llc/pigeon#26]
This is intended to start a conversation about changes need to handle multiple environments or multiple topics.

With coming changes to APNS I suspect it will be more common to deliver notifications to several topics in the same application. For example pushing to both production and sandbox or pushing to two separate topics.

Currently the certificate to use is specified in the config but it should be possible to specify a push topic and load a map between certificates and topics instead

```
  %{ "com.exmaple.topic1":"/path/to/cert1.p12", "com.exmaple.topic2":"/path/to/cert2.p12"}
```

and the add the option to push against a Topic. At some point Apple will move over to JWT which should simplify this dramatically.

This probably requires a separate pool per topic.

**Comment by hpopp** [comment]
I agree that multiple configurations are needed, though the config needs to remain dead simple for the majority of use cases. Perhaps define another pigeon config block scoped to the push topic?

Something like:
```elixir
# Optional default config. If a topic hasn't been mapped it uses this.
config :pigeon,
  apns_mode: :dev,
  apns_cert: "default_cert.pem",
  apns_key: "default_key.pem"

config :pigeon, "com.example.topic1",
  apns_mode: :dev,
  apns_cert: "cert1.pem",
  apns_key: "key1.pem"

config :pigeon, "com.example.topic2",
  apns_mode: :prod,
  apns_cert: "cert2.pem",
  apns_key: "key2.pem",
  apns_2197: true
```

This would allow a user to specify `:dev`/`:prod` and port `2197` on a per-topic basis.

**Comment by hpopp** [comment]
Hah, I guess mix configs don't play that nice with string keys. It'd have to look more like

```elixir
config :pigeon, :default
  apns_mode: :dev,
  apns_cert: "default_cert.pem",
  apns_key: "default_key.pem"

config :pigeon, :"com.example.topic1",
  apns_mode: :dev,
  apns_cert: "cert1.pem",
  apns_key: "key1.pem"

config :pigeon, :"com.example.topic2",
  apns_mode: :prod,
  apns_cert: "cert2.pem",
  apns_key: "key2.pem",
  apns_2197: true
```

**Comment by hpopp** [comment]
Lot of functionality in 6af6665a3e54c7354e6d1dff66248c4a042a8f07 on the `multiconn` branch. Poolboy has been added with dynamic creation/deletion of worker pools, configured like
```
Pigeon.APNS.start_connection(pool_name, :dev/:prod, "cert.pem", "key.pem")
```

The cert and key files can be specified, just like in the config, as either a file name, `{:pigeon, "filename.pem"}`, or the full text. It sets up 5 workers by default. Configuring pool size has yet to be implemented.

`Pigeon.APNS.push` now takes a notification and an optional keyword list or arguments.
E.g. sending to a specific pool with a callback:
```
Pigeon.APNS.push(notification, on_response: callback_fn, name: pool_name)
```

I'll merge once custom connections can be specified in the config and I'm happy with the API changes. Hopefully this will be the last API-breaking change.

**Comment by idyll** [comment]
Sorry for the delay in my response. I took some time off before heading to ElixirConf this week.

Would you consider changes to support reading the P12 directly in addition to the pems? If so can provide a PR with code to do it. If I remember correctly I need to call out to the erlang crypto to parse the P12. I did some PKCS#7 stuff recently so parsing and decrypting the P12 should be straighforwards.

In addition I could add support for password protected P12 and key.pems. This way you can put the description key in an ENV var and it feels a little safer to toss into version control (for projects that need only one cert.)

**Comment by hpopp** [comment]
That would be fantastic!

**Comment by hpopp** [comment]
Finalized the config API in https://github.com/codedge-llc/pigeon/commit/38d0ef6bb88fc0e32373234cb248b313b47ec473.

I realize now that poolboy is useless for GenServer casts. I need to remove it and perhaps manually create some sort of round-robin system for casting to workers.

**Comment by zepplock** [comment]
@hpopp Have you considered using `GenStage`?

**Comment by hpopp** [comment]
I'll take a look. I remember reading about it some time ago but have yet to use it in a project.

**Comment by hpopp** [comment]
Closed with #40
